src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1185,8 +1189,9 @@ private string GetPreferredExtension(DocumentId documentId, SourceCodeKind sourc
 
         public override IVsHierarchy GetHierarchy(ProjectId projectId)
         {
-            // This doesn't take a lock since _projectToHierarchyMap is immutable
-            return _projectToHierarchyMap.GetValueOrDefault(projectId, defaultValue: null);
+            // This doesn't take a lock since _projectToGuidMap is immutable
+            return _projectToGuidMap.TryGetValue(projectId, out var projectGuid) &&
+                   ErrorHandler.Succeeded(_vsSolution.GetProjectOfGuid(ref projectGuid, out var hierarchy)) ? hierarchy : null;
```

Comment:
This has now affinitized this method; people using IVsHierarchy of course can't use the result, but if somebody is calling it on a background thread and then doing object identity comparisons that's something they *could* have done.

Reply:
I'll add 

```C#
            _foregroundObject.AssertIsForeground();
```

so we get a good error if that happens.